### PR TITLE
Reduce the number of synchronizations produced.

### PR DESCRIPTION
**Pull request by math-fehr**
This PR aims to reduce the number of synchronizations produced by the compiler :

1. Added functions to be able to insert __syncwarp in the schedule tree and to use it with all versions of CUDA  (added a definition of __syncwarp when the compiler version is below 9).
2. Memorize the dependences in scop instead of recomputing it later. Theses dependences are necessary to know when synchronizations are needed.
3. The function findBestSync compute the lowest synchronization necessary between two schedule trees. The function is used by insertBestSyncInSeq to reduce the number of __syncthreads and __syncwarp inserted in a sequence during thread mapping using dynamic programming. As for now, the number of __syncthreads are minimized first, then the number of __syncwarp.
4. When memory reads/writes are inserted, remove the synchronizations between two reads, or two writes.

### CODE REVIEW

File: tc/core/libraries.h [hunk 1]
```diff
@@ -43,6 +43,13 @@ constexpr auto defines = R"C(
 #define inf __longlong_as_double(0x7ff0000000000000LL)
 )C";
 
+constexpr auto warpSyncFunctions = R"C(
```
Comment (ftynse) [inline tc/core/libraries.h]: Let's have a comment above saying that "before cuda 9, warps are implicitly synchronized so syncwarp is a noop"

File: tc/core/polyhedral/scop.h [hunk 2]
```diff
@@ -229,6 +245,20 @@ struct Scop {
     return true;
   }
 
+  static bool isWarpSyncId(isl::id id) {
+    auto name = id.get_name();
```
Comment (ftynse) [inline tc/core/polyhedral/scop.h]: I would abstract this and `isSyncId` away to a function that takes an extra string argument of expected id prefix.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 3]
```diff
@@ -649,6 +652,9 @@ std::unique_ptr<MappedScop> MappedScop::makeWithOuterBlockInnerThreadStrategy(
     scop->specializeToContext();
   }
 
+  // 9. Insert mapping context
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: This change is independent of dependence computation and can be moved to a separate commit.  That would provide space to motivate the change.

Also, let's please change the numbering in the comment and use 5 here and 6+ below.

File: tc/core/polyhedral/scop.cc [hunk 4]
```diff
@@ -371,30 +371,19 @@ isl::union_map computeDependences(
   return flow.get_may_dependence();
 }
 
-// Do the simplest possible dependence analysis.
 // Live-range reordering needs tagged access relations to be available.
```
Comment (ftynse) [inline tc/core/polyhedral/scop.cc]: This comment is also related to the dependence analysis and should be removed

File: tc/core/polyhedral/scop.h [hunk 5]
```diff
@@ -437,6 +438,13 @@ struct Scop {
       isl::schedule_constraints constraints,
       const SchedulerOptionsView& schedulerOptions);
 
+ public:
+  // Do the simplest possible dependence analysis.
+  // Compute all RAW, WAR, and WAW dependences, and save them in dependences.
+  void computeAllDependences();
+  // Save the RAW, WAR, and WAW dependences
```
Comment (ftynse) [inline tc/core/polyhedral/scop.h]: Saying "save" sounds like a function, but it's a member.  Let's just say "RAW, WAR and WAW dependences"

File: tc/core/polyhedral/scop.h [hunk 6]
```diff
@@ -437,6 +438,13 @@ struct Scop {
...
+  isl::union_map dependences;
```
Comment (ftynse) [inline tc/core/polyhedral/scop.h]: The styling policy is to put members below all functions.  In this case, let's put this below read/write map in line 484.

File: tc/core/polyhedral/cuda/mapped_scop.h [hunk 7]
```diff
@@ -155,6 +155,15 @@ class MappedScop {
   // The remaining parts, if any, are no longer considered for replacement
   // by a library call.
   detail::ScheduleTree* separateReduction(detail::ScheduleTree* band);
+
+  // Find best thread sync between st1 and st2 when st2 is scheduled after
+  // st1.
+  // 0 indicates no sync are needed,
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.h]: can we have an `enum` with meaningful names instead?

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 8]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+    isl::aff aff(isl::aff::param_on_domain_space(space, id));
+    formula = formula * mappingIds[i].second + aff;
+  }
+
+  return formula;
+}
+
+// Return the constraints ensuring that the points with parameters
+// [t0,t1,t2] and [t0',t1',t2'] are in the same warp.
+// (where t0 is "t0" + suffix1 and t0' is "t0" + suffix2)
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: Here and in `getThreadIndexFormula`, it would be helpful to describe the shape of the aff/set in the comment

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 9]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+  auto contextSt = scop_->scheduleRoot()->children()[0];
+  auto contextElem = contextSt->elemAs<detail::ScheduleTreeElemContext>();
+  CHECK(nullptr != contextElem);
+  domainAndContext = domainAndContext.intersect_params(contextElem->context_);
+
+  // The domain of both schedule trees filtered by mapping filters,
+  // and then modified to have different threads and blocks names.
+  auto domain1 = intersectMappingFilterParams(st1, domainAndContext);
+  auto domain2 = intersectMappingFilterParams(st2, domainAndContext);
+  domain1 = modifyMappingNames(domain1, "_1");
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: Let's extract `"_1"` and `"_2"` into local constants with meaningful names

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 10]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+    {TZ, TZ.mappingSize(block)}};
+
+  auto formula1 = getThreadIdxFormula(space, block, suffix1);
+  auto formula2 = getThreadIdxFormula(space, block, suffix2);
+
+  return (isl::aff_set((formula1 / warpSize).floor()) == (formula2 / warpSize).floor());
+  }
 } // namespace
 
+int MappedScop::findBestSync(
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: This function assumes it is called above thread mapping (and below block mapping).  I would document this assumption.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 11]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
   }
   return false;
 }
+
+// Intersect the union set with all the mapping
+// filters params in the given schedule tree
+isl::union_set intersectMappingFilterParams(
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: I don't quite understand what this function returns if there are multiple _divergent_ thread mappings in different children, for example one branch is mapped to (32,4) threads and another to (7,1) threads.  Some reasoning in the comment would be helpful.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 12]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+      auto ii = i % nChildren;
+      if(ik == ii) {
+        bestSync[i][k] = 0;
+      } else {
+        bestSync[i][k] = findBestSync(children[ii], children[ik]);
+      }
+    }
+  }
+
+  // Get the worse synchronization needed in the sequence
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: "worse" -> "strictest" ?

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 13]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+
+  // The optimal number of block sync and thread sync needed to
+  // have the sequence children[i], ..., children[j] correctly
+  // synchronized
+  std::vector<std::vector<std::pair<int, int>>> optimalValue(
+      size, std::vector<std::pair<int, int>>(range, {-1, -1}));
+
+  // An optimal position for doing a synchronization between children[i]
+  // and children[j]. The second member indicates which synchronization
+  // should be made. This should be use recursively to get the optimal
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: use -> used

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 14]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+  }
+
+  // The optimal number of block sync and thread sync needed to
+  // have the sequence children[i], ..., children[j] correctly
+  // synchronized
+  std::vector<std::vector<std::pair<int, int>>> optimalValue(
+      size, std::vector<std::pair<int, int>>(range, {-1, -1}));
+
+  // An optimal position for doing a synchronization between children[i]
+  // and children[j]. The second member indicates which synchronization
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: And what is the first member?

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 15]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+  // should be made. This should be use recursively to get the optimal
+  // synchronization. If the second member is equal to 0, this means that
+  // no synchronization is needed.
+  std::vector<std::vector<std::pair<int, int>>> optimalSyncPosition(
+      size, std::vector<std::pair<int, int>>(range, {-1, -1}));
+  for (size_t i = 0; i < size; ++i) {
+    optimalValue[i][0] = {0, 0};
+    optimalSyncPosition[i][0] = {0, 0};
+  }
+  for (size_t k = 1; k < range; ++k) {
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: I wonder if it would be possible to extract the DP into a separate function (does not seem to need anything other than the input matrix) and write a test for it comparing against a brute-force implementation on a small size?

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 16]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+    if(optimalSyncPosition[i][k].second != 0) {
+      auto separation = optimalSyncPosition[i][k].first;
+      stack.push_back({i, separation - i});
+      stack.push_back({separation + 1, i + k - separation - 1});
+      synchronizations.push_back(optimalSyncPosition[i][k]);
+    }
+  }
+
+  // Insert all the synchronizations
+  std::sort(synchronizations.begin(), synchronizations.end());
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: You may sort with `std::greater` to avoid iterating in the inverse order.

**Comment by ftynse** [comment]
typo in bb86ece2754a1a91ed738efc678d9d688a00c7b1 comment "choose" -> "chooses"
typos in 199e26cbe1d223d703b26a3b8723289408b9f836 comment "read" -> "reads", "write" -> "writes", "points" -> "point"

7a771d9794a266435d2f229b23a60a586956ed7d can be squashed into ae75cb3c9051a77ec8bbd8c7da2ae99bf6770b05

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 17]
```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
...
+  if (auto mappingFilter = st->elemAs<detail::ScheduleTreeElemMappingFilter>()) {
+    us = us.intersect(mappingFilter->filter_);
+  } else if (auto filter = st->elemAs<detail::ScheduleTreeElemFilter>()) {
+    us = us.intersect(filter->filter_);
+  }
+
+  auto children = st->children();
+  auto nChildren = children.size();
+  if (nChildren == 1) {
+    us = intersectMappingFilterParams(children[0], us);
```
[the remark this answers is not in the document]
Comment (math-fehr) [reply tc/core/polyhedral/cuda/mapped_scop.cc]: That was a mistake, I shouldn't have taken the union with the superset, but only with the subsets.

File: tc/core/polyhedral/cuda/mapped_scop.cc (hunk 11 above)
Comment (math-fehr) [reply tc/core/polyhedral/cuda/mapped_scop.cc]: This function add the constraints contained in the mapping filters to the active domain points in the mapping filter node.
If we have a domain { S0[i0] : 0 <= i0 < 42;   S1[i1] : 0 <= i1 < 27 },
followed by a sequence splitting S0 and S1 with filters, and a mapping filter in each branch mapping respectively t0 to i0 on 4 threads and t0 to i1 on 3 threads.
The result would be { S0[i0] : 0 <= i0 < 42 and t0 = i0 mod 4 ;  S1[i1] : 0 <= i1 < 27 and t0 = i1 mod 3 }
